inst/include/newton_solver.h

```diff
@@ -0,0 +1,115 @@
+// Copyright 2023 'finalsize' authors. See repository licence in LICENSE.md.
+#ifndef INST_INCLUDE_NEWTON_SOLVER_H_
+#define INST_INCLUDE_NEWTON_SOLVER_H_
+
+// clang-format off
+#include <Rcpp.h>
+#include <RcppEigen.h>
+// clang-format on
+
+// add to namespace finalsize
+namespace finalsize {
+
+/// function for newton solver
+// taken from Edwin van Leeuwen at
+// https://gitlab.com/epidemics-r/code_snippets/feature/newton_solver/include/finalsize.hpp
+
+/// @brief Solve an equation using the Newton method with tolerance checking
+/// @param contact_matrix Social contact matrix. Entry \eqn{mm_{ij}} gives
+/// the average number of contacts in group \eqn{i} reported by participants
+/// in group \eqn{j}.
+/// @param demography_vector Demography vector. Entry \eqn{pp_{i}} gives
+/// proportion of total population in group \eqn{i}. Normalised if needed.
+/// @param susceptibility A matrix giving the susceptibility of individuals
+/// in demographic group \eqn{i} and risk group \eqn{j}.
+/// @param iterations Number of solver iterations. Defaults to 10,000.
+/// @param tolerance Solver error tolerance. Solving ends when the error
+/// drops below this tolerance. Defaults to set `1e-6`. Larger tolerance
+/// values are likely to lead to inaccurate final size estimates.
+/// @return A vector of final sizes.
+inline Eigen::ArrayXd solve_final_size_newton(
+    const Eigen::MatrixXd &contact_matrix,
+    const Eigen::VectorXd &demography_vector,
+    const Eigen::VectorXd &susceptibility, const int iterations = 10000,
+    const double tolerance = 1e-6) {
+  // count number of demography groups
+  int nDim = demography_vector.size();
+
+  Eigen::ArrayXi zeros(nDim);
+  zeros.fill(0);
+
+  Eigen::ArrayXd epi_final_size(nDim);  // prev in settings struct
+  epi_final_size.fill(0.5);
+
+  Eigen::MatrixXd contact_matrix_ = contact_matrix;
+  for (int i = 0; i < contact_matrix.rows(); ++i) {
+    // Check if value should be 0 for (limited) performance increase
+    if (demography_vector(i) == 0 || susceptibility(i) == 0 ||
+        contact_matrix.row(i).sum() == 0) {
+      zeros[i] = 1;
+      epi_final_size[i] = 0;
+    }
+    for (int j = 0; j < contact_matrix.cols(); ++j) {
+      if (zeros[j] == 1) {
+        contact_matrix_(i, j) = 0;
+      } else {
+        // Scale contacts appropriately
+        // Could add transmissibility (j)?
+        contact_matrix_(i, j) =
+            susceptibility(i) * contact_matrix(i, j) * demography_vector(j);
+      }
+    }
+  }
```

Comment:
On my machine with the following example:

https://github.com/epiverse-trace/finalsize/blob/2d7c68177f718e917a8853a6d7415f5fb6524318/tests/testthat/test-solver_speedups.R#L3-L44

|     | min     | lq       | mean     | median   | uq       | max      | neval |
|-----|---------|----------|----------|----------|----------|----------|-------|
| old | 387.414 | 448.9645 | 549.276  | 480.014  | 546.0495 | 151736.6 | 10000 |
| new | 381.732 | 449.15   | 544.8088 | 483.8075 | 558.802  | 6673.177 | 10000 |

Reply:
Also, if we end up removing `zeros`, this entire step should probably moved to the `final_size()` function so you don't have to duplicate it in both solvers.

Reply:
I've also removed `zeros` overall, and moved the contact matrix scaling to the main function. Instead, I'm checking whether the iterative solver should return a zero final size when the (scaled) contact matrix rowsums are zero. More such conditions could be added in case the contact matrix and demography vector are out of sync (i.e., groups have contacts but reported empty).

For now, tests pass for the numerical error correction even without using `zeros` in the solver files.